Golang implementation of Time Series Merge

mrtracy opened the pull request:
This commit creates a Golang implementation of the logic to merge
InternalTimeSeriesData messages together; this has also been implemented in C++
as part of our RocksDB custom merge operator.

The Go implementation is primarily intended for testing purposes - it provides
the ability to double-check the logic present in the C++ engine as part of high
level, model based testing which is currently in development. The Golang
implementation is run through the exact same merge tests as the engine
implementation.

As part of this commit, I have also repaired a bug in the C++ engine merge logic
; previously, it was not properly sorting and accumulating the samples from the
first merge operation on a given key, instead just accepting the collection
exactly as it was provided.

Commit:
Fix Time Series Merge

This commit repairs a bug in the C++ engine merge logic ; previously, it was not
properly sorting and accumulating the samples from the first merge operation on
a given key, instead just accepting the collection exactly as it was provided.

This commit also exports the C++ merging logic for Time Series as the
"MergeInternalTimeSeriesData" method in the engine package. This method is being
exported for usage in higher-level time series tests which are forthcoming.

petermattis commented on storage/engine/db.cc:
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
+    // Sort values in the ts value. 
+    std::sort(val_ts.mutable_samples()->pointer_begin(),
+            val_ts.mutable_samples()->pointer_end(),
+            TimeSeriesSampleOrdering);
```
Looks like these two lines are mis-indented.

petermattis replied:
Not part of this change, but you should use a functor object instead of a function pointer for `TimeSeriesSampleOrdering`. In my past experience functor objects inline better (might not be true of recent compilers). Something like:

```
struct TimeSeriesSampleOrdering {
  bool operator()(const cockroach::proto::InternalTimeSeriesSample* a, const cockroach::proto::InternalTimeSeriesSample* b) const {
    return a->offset() < b->offset();
  }
};
```

petermattis commented on storage/engine/db.cc:
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
+         end = val_ts.samples().end();
+
+    // Loop until samples have been exhausted.
+    while(front != end) {
```
`s/while(/while (/g`

mrtracy replied:
Done

petermattis commented on storage/engine/db.cc:
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
+
+    // Merge sample values of left and right into new_ts.
+    auto front = val_ts.samples().begin(),
+         end = val_ts.samples().end();
```
In C++ I prefer to see 1 variable definition per line and to avoid using commas to separate multiple variable declarations. Pretty sure this is part of the Google C++ coding guidelines which we generally follow.

tnachen replied:
Yes Google C++ style states so too.

mrtracy replied:
Done

petermattis commented on storage/engine/db.cc:
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
+        ns->set_offset(front->offset());
+        while (front != end && front->offset() == ns->offset()) {
+            AccumulateTimeSeriesSamples(ns, *front);
+            front++;
```
With C++ iterators you should prefer pre-increment to post-increment: `s/front++/++front/g`

mrtracy replied:
Any reason for that?  Style guide, or an implementation advantage?

petermattis replied:
It's an implementation advantage. The post-increment requires a copy. The pre-increment does not. See the difference between https://code.google.com/p/cpp-btree/source/browse/btree.h#812 (pre-increment) and https://code.google.com/p/cpp-btree/source/browse/btree.h#820 (post-increment).

petermattis commented on storage/engine/merge_test.go:
```diff
@@ -353,20 +365,51 @@ func TestGoMerge(t *testing.T) {
+		return nil
+	}
+	var mvccValue proto.MVCCMetadata
+	gogoproto.Unmarshal(b, &mvccValue)
```
Doesn't `gogoproto.Unmarshal` return an error? You should be checking it.

mrtracy replied:
Done

petermattis commented on proto/internal.go:
```diff
@@ -87,3 +91,190 @@ func InternalTimeSeriesDataFromValue(value *Value) (*InternalTimeSeriesData, err
+
+// AccumulateFrom accumulates values from the supplied InternalTimeSeriesSample
+// into the existing values of the destination InternalTimeSeriesSample.
+func (dest *InternalTimeSeriesSample) AccumulateFrom(src *InternalTimeSeriesSample) {
```
Any reason for this to be part of the `proto` package. If it is only used by `storage/engine/merge_test.go`, I'd prefer for it to be private to the `storage/engine` package.

mrtracy replied:
It's not only going to be used there, it's also going to support testing in the upcoming /ts module.  That's the primary reason for this PR.

petermattis commented:
LGTM

spencerkimball commented on proto/internal.go:
```diff
@@ -87,3 +91,190 @@ func InternalTimeSeriesDataFromValue(value *Value) (*InternalTimeSeriesData, err
+			return defaultVal
+		}
+		var (
+			destMax = valOrDefault(dest.IntMax, dest.GetIntSum())
```
Looking at this, I'm wondering whether you think it might be overkill to support both Ints and Floats. I know we discussed this before but I don't recall the convincing argument that we support both. I'd be tempted to just support Float64. I'm worried about all of these allocations. hmm.

spencerkimball commented:
I'm not a big fan of implementing the same logic twice. If we only need the go logic in order to do testing, let's just create a cgo method which allows the C++ merging code to be called from Go. You can just pass the cgo method the serialized protobufs and deserialize on the C++ side, then run through the C++ merge, serialize again and then deserialize the result on the Go side.

Maybe I'm missing something...

mrtracy commented:
The C++ Merging code can already be called from go (it's needed for batch),
but it's not being exported from the engine module.  I need this for the
model-based testing i'm using in the /ts module, for which I haven't yet
sent out a PR.

I made the decision that re-implementing the logic in Go would be
beneficlal, as it is a fairly complicated transformation.  The real
implementation is an optimized C++ implementation which is called in the
context of a RocksDB Merge Operator, and so it's not especially
straightforward to understand.  I feel that this is a solid way to double
check the implementation in more complicated situations.

If you feel strongly about having only one implementation, I can export the
rename and export the goMerge method from the engine layer.

On Fri, Apr 3, 2015 at 4:47 PM, Spencer Kimball <notifications@github.com>
wrote:

> I'm not a big fan of implementing the same logic twice. If we only need
> the go logic in order to do testing, let's just create a cgo method which
> allows the C++ merging code to be called from Go. You can just pass the cgo
> method the serialized protobufs and deserialize on the C++ side, then run
> through the C++ merge, serialize again and then deserialize the result on
> the Go side.
>
> Maybe I'm missing something...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/518#issuecomment-89429722>.
>

spencerkimball commented:
I actually do feel pretty strongly that we shouldn't duplicate this code. It's great that it allowed you to find a bug, but I think ultimately having two implementations will lead to more confusion and maintenance than it's worth. If you want to use it for model based testing in the /ts module, let's just provide some exported methods from engine which call into the C++.

mrtracy commented:
I have removed the go implementation (which significantly changes the purpose of this commit) and replaced it with an exported helper method in the engine, which directly provides for the merging of InternalTimeSeriesData objects.

spencerkimball commented on storage/engine/db.cc:
```diff
@@ -492,6 +492,54 @@ bool MergeTimeSeriesValues(cockroach::proto::Value *left, const cockroach::proto
     return true;
 }
 
+// ConsolidateTimeSeriesValue processes a single value which contain
```
s/contain/contains/

spencerkimball commented:
LGTM
